Re-attempting rosserial for VEX Cortex

CanyonTurtle opened the pull request:
### Overview 

The added package, `rosserial_vex_cortex`, allows the user to use rosserial on the [VEX Cortex](https://www.vexrobotics.com/276-2194.html), ontop of the [PROS kernel](https://pros.cs.purdue.edu/cortex/index.html).

This is a re-attempt at #378, based on feedback from @romainreignier and @tfoote .
This time the added project is self-contained, has no unneeded PROS project files, and allows the user to generate PROS projects and install the rosserial libraries automatically (see README.md).

### Limitations
Only tested on Linux 18.04LTS + ros melodic, relies on bash script.

CanyonTurtle committed:
initial commit

CanyonTurtle committed:
reorganized dir to provide project building script

CanyonTurtle committed:
readme change

CanyonTurtle committed:
readme change

tfoote commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,62 @@
+3. PROS - [install PROS toolchain](https://github.com/purduesigbots/pros/releases/tag/2.12.1)
+
+# Setup
+Stup a ROS workspace and build rosserial packages (including rosserial_vex_cortex) from source:
```
typo s/Stup/Setup/

tfoote commented on rosserial_vex_cortex/src/rosserial_vex_cortex/make_libraries.py:
```diff
@@ -0,0 +1,91 @@
+rospack = rospkg.RosPack()
+
+# copy ros_lib stuff in
+rosserial_cortex_dir = rospack.get_path(THIS_PACKAGE)
```
It looks like this script is a duplicate of the one in rosserial_arduino. Except for the THIS_PACKAGE variable. It might be a good candidate for parameterizing and deduplicating in the future.

tfoote commented on rosserial_vex_cortex/src/rosserial_vex_cortex/make_libraries.py:
```diff
@@ -0,0 +1,91 @@
+
+import rospkg
+import rosserial_client
+from rosserial_client.make_library import *
```
Importing * is a bad practice. I know it's done in the other version, but explicitly calling the things out make it much easier to debug.

tfoote commented on rosserial_vex_cortex/src/ros_lib/ros.h:
```diff
@@ -0,0 +1,70 @@
+
+namespace ros
+{
+// #if defined(__AVR_ATmega8__) or defined(__AVR_ATmega168__)
```
You should clean out these commented values that are from the other verison.

Also above.

tfoote commented on rosserial_vex_cortex/src/rosserial_vex_cortex/SerialClient.py:
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python
```
Can we not use this file from another instance/location? I don't see any local modifications.

tfoote commented on rosserial_vex_cortex/src/ros_lib/logger.h:
```diff
@@ -0,0 +1,51 @@
+// 1 for yes
+#define DEBUG_MODE 0
+
+// apply a macro to all args of a variadic macro, used for logg
```
These old commented versions should be cleaned up.

tfoote commented on rosserial_vex_cortex/src/ros_lib/logger.h:
```diff
@@ -0,0 +1,51 @@
+// will print to uart2. 
+// Use screen to access uart2 on the terminal with a serial-usb adapter+cable
+// on device USB0 at 115200 baud: screen /dev/ttyUSB0 115200
+#define loggy(fmtstr, ...) fprintf((uart2), "[%d]: "       \
```
The names of these functions should be cleaned up too.

tfoote commented on rosserial_vex_cortex/src/ros_lib/logger.h:
```diff
@@ -0,0 +1,51 @@
+#ifndef __CANNON_LOGGER_H__
+#define __CANNON_LOGGER_H__
```
Include guards should be `PACKAGE_NAME_FILE_NAME_H` format. http://wiki.ros.org/CppStyleGuide

tfoote reviewed (commented):
Overall this is much cleaner. I have a bunch of smaller style and cleanup suggestions.  There's also a few places that look like they're ripe for adding/improving reuse. Those might be worth deferring for another PR.

Also depending on the thoughts of maintainers we could work in a branch in parallel with master to be less disruptive as we iterate on Vex support.

CanyonTurtle committed:
removed more
unnecessary parts

CanyonTurtle committed:
renamed logger functions to be more specific.

CanyonTurtle committed:
logger name changes

CanyonTurtle committed:
added to the readme

CanyonTurtle commented:
@tfoote, Thanks for the suggestions. I have removed the unnecessary `serial-node` script from this subproject and cleaned up the function names, packages, and documentation.

tfoote commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,66 @@
+# Requirements
+1. Linux (Only tested on Ubuntu 18.04LTS)
+2. ROS (Only tested on ROS Melodic) - [installation guide](http://wiki.ros.org/melodic/Installation/Source).
+3. PROS - [installation guide](tps://github.com/purduesigbots/pros/releases/tag/2.12.1)
```
Broken link and it says installation gude, but points to the tarball.

tfoote commented on rosserial_vex_cortex/src/ros_lib/helloworld.cpp:
```diff
@@ -0,0 +1,42 @@
+#include <ros.h>
+#include <std_msgs/String.h>
+
+// run continuously by the setup function, publishes the time at 50hz.
```
This comment is out of date.

tfoote commented on rosserial_vex_cortex/src/ros_lib/logger.h:
```diff
@@ -0,0 +1,39 @@
+
+#include "API.h"
+
+// Use screen to access uart2 on the terminal with a serial-usb adapter+cable
```
Documentation of the debug port would be good to add as a block in the README.

romainreignier replied:
:+1:

tfoote reviewed (commented):
This looks much better. Just a few more comments.

CanyonTurtle committed:
readme diagram

CanyonTurtle committed:
image to specify uart configuration

romainreignier commented on rosserial_vex_cortex/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+project(rosserial_vex_cortex)
+
+find_package(catkin REQUIRED COMPONENTS
+  message_generation
```
No message is generated so this dependency is not needed.

romainreignier commented on rosserial_vex_cortex/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# put the names of custom service files here
+# add_service_files(FILES
+# 
+# )
```
The same apply for this commented lines, they are not needed.

romainreignier commented on rosserial_vex_cortex/CMakeLists.txt:
```diff
@@ -0,0 +1,40 @@
+# generate_messages()
+
+catkin_package(
+  CATKIN_DEPENDS message_runtime
```
And also this dependency is not used

romainreignier commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,88 @@
+# rosserial for VEX Cortex
+
+This package contains everything needed allow you to run rosserial on the [VEX Cortex](https://www.vexrobotics.com/276-2194.html), on the [PROS Kernel](https://pros.cs.purdue.edu/cortex/index.html).
```
s/needed allow/needed to allow/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+catkin_make install # this will generate folders in the workspace that contain executable scripts. 
+```
+
+# Usage Example(Linux)
````
s/Example(Linux)/Example (Linux)/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+source /opt/ros/melodic/setup.bash # or replace melodic with your corresponding ROS version name
+roscore 
+```
+this will start ROS on the host linux machine.
````
s/this/This/
s/linux/Linux/

romainreignier commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,88 @@
+This will show you how to run the "hello world" example PROS project.
+
+### Step 1 : Run the ROS instance
+open a terminal window and run:
```
s/open/Open/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+```
+this will start ROS on the host linux machine.
+
+### Step 2: get the VEX Cortex to talk
````
s/get/Get/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+```
+
+### Step 4: Display a topic in a terminal (technically optional)
+open another terminal window and run:
````
s/open/Open/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+source /opt/ros/melodic/setup.bash # or replace melodic with your corresponding ROS version name
+rostopic echo chatter
+```
+this will display any messages that come through on the "chatter" topic.
````
s/this/This/

romainreignier commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,88 @@
+
+# Debugging on Cortex
+PROS does not provide a debugger, but it does provide printing.
+use vexroslog(char* out, ...) just like you would use fprintf from the PROS API:
```
s/use/Use/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+```bash
+vexroslog("hello, my favorite number is %d", 3);
+```
+remember to include this header in your code for logging!
````
s/remember/Remember/

romainreignier commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,88 @@
+
+`include/logger.h`, you can change which serial output to use (stdin/stdout, uart1, or uart2) for debugging.
+
+to view serial output from UART2, (instead of running `pros terminal`, which only works for stdout), use screen:
```
s/to/To/

romainreignier commented on rosserial_vex_cortex/README.md:
````diff
@@ -0,0 +1,88 @@
+# to install screen: sudo apt-get install screen
+screen /dev/ttyUSB0
+```
+note: this requires a [USB-serial adapter](https://www.adafruit.com/product/954) for your computer, and it needs to be plugged in correctly.
````
s/note/Note/

romainreignier commented on rosserial_vex_cortex/README.md:
```diff
@@ -0,0 +1,88 @@
+to view the UART 
+
+# Limitations
+Global scope causes segmentation faults. Do not try to use global scope. ROS-related objects in the global scope causes segmentation faults. The cause of this is still unknown, but it may have something to do with the implementation of the global memory segment.
```
The first and third sentences are quite similar (only "ROS-related objects" differ). Maybe a reworking would help to clarify.

Maybe something like:
Global scope variables causes segmentation faults. Unlike in the other `rosserial` examples, avoid ROS objects in the global scope.

romainreignier commented on rosserial_vex_cortex/package.xml:
```diff
@@ -0,0 +1,22 @@
+  <run_depend>rospy</run_depend>
+  <run_depend>rosserial_msgs</run_depend>
+  <run_depend>rosserial_client</run_depend>
+  <run_depend>message_runtime</run_depend>
```
As mentioned earlier, does not seems useful.

romainreignier commented on rosserial_vex_cortex/src/ros_lib/CortexHardware.h:
```diff
@@ -0,0 +1,76 @@
+    // any initialization code necessary to use the serial port
+    void init() {
+      // uart2 is for debugging, while stdout is initialized automatically.
+      usartInit(uart2, 115200, SERIAL_8N1);
```
Usually in `rosserial`, the default baudrate is 57600.

And the hardware class can offer a method to change it, see `rosserial_arduino`: https://github.com/ros-drivers/rosserial/blob/fe1e233f39d56931b5e223c5e53a2d9c2f9fb664/rosserial_arduino/src/ros_lib/ArduinoHardware.h#L88-L90

romainreignier commented on rosserial_vex_cortex/src/ros_lib/helloworld.cpp:
```diff
@@ -0,0 +1,43 @@
+/*
+ * rosserial Publisher Example for VEX Cortex
```
If possible, move the examples to a separate folder like in `rosserial_arduino`: https://github.com/ros-drivers/rosserial/tree/fe1e233f39d56931b5e223c5e53a2d9c2f9fb664/rosserial_arduino/src/ros_lib/examples

romainreignier commented on rosserial_vex_cortex/src/ros_lib/ros.h:
```diff
@@ -0,0 +1,46 @@
+#include "CortexHardware.h"
+
+namespace ros {
+  typedef NodeHandle_<CortexHardware, 2, 2, 128, 128> NodeHandle;
```
The first two template arguments are the max number of subscribers and publishers;
https://github.com/ros-drivers/rosserial/blob/fe1e233f39d56931b5e223c5e53a2d9c2f9fb664/rosserial_client/src/ros_lib/ros/node_handle.h#L99-L104

Are you really limiting the users of the default version of `rosserial_vex_cortex` to use a maximum of **two** subscribers and **two** publishers?

romainreignier reviewed (changes requested):
This new PR seems in a good shape to me.

Nonetheless, I have added a few minor comments, mostly cosmetic fixes in the README.

But I have a concern about the number of MAX_SUBSCRIBERS and MAX_PUBLISHERS. 2 seems very low to me.

CanyonTurtle committed:
adjusted readme, documentation, and slimmed project structure

CanyonTurtle commented:
@tfoote @romainreignier,

Thanks for the continued feedback. I was able to slim the package even more, and I fleshed out the documentation for setting up and configuring the serial connection, and touched it up overall. I also followed every suggestion here (except the `setBaud` hardware member function @romainreignier  - I think the serial port must be configured in a special part of the VEX Cortex lifecycle, so I neglected this change. See `init.cpp`).

 @romainreignier I also reset the publisher and subscriber counts to their defaults. The numbers being so low last time was leftover from my testing. Thanks for the find!

One topic of discussion: should I create a new branch for continued development of this package in coming weeks, or should I make future potential PRs on `jade-devel`?

romainreignier commented on rosserial_vex_cortex/CHANGELOG.rst:
```diff
@@ -1,6 +1,11 @@
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 Changelog for package rosserial_vex_cortex
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+0.1.1
```
I think that all `rosserial` packages should have the same version number so it should be 0.7.7. But I may be wrong. See with @mikepurvis

CanyonTurtle replied:
thanks, about to re-commit because my commit broke the building (oops).

romainreignier reviewed (commented):
About the branch, I am not sure, but if you don't touch any other `rosserial` packages, it should be ok to stay with `jade-devel`.

But for these kind of questions, maybe @mikepurvis can answer.

CanyonTurtle committed:
re-added some necessary infrastructure

CanyonTurtle commented:
Oop, just re-added a couple files and dependencies, I broke the vex cortex build process with the last commit. This one is tested and working.

CanyonTurtle commented:
@mikepurvis, what do you think of this PR? I think it is in good condition to accept. This version of rosserial_vex_cortex is self-contained and documented in-package, and has been closely reviewed by @tfoote and @romainreignier.

As I alluded to before, I intend to add further documentation and some additional examples to this package. For that purpose, do you think I should branch off, or stay on `jade-devel`?

CanyonTurtle committed:
Added documentation about how to prevent crashes.

CanyonTurtle committed:
added demos for using a joystick or a phone to control robot

CanyonTurtle commented:
Note @mikepurvis : Right now, I am continually developing this package, but this version is stable. To facilitate continued development, I will leave this PR as is, for reviewing purposes, and I will branch off of my own fork for the continued changes I am working on for a separate PR later.

CanyonTurtle commented:
@mikeferguson @mikepurvis,

Would you please tell me what you think of this additional hardware support? I am looking to get this through because I have commits that build off of this. I also have VEX V5 support, which is a brand-new platform that will be used by thousands of VEX Robotics teams in the coming few years.

mikepurvis commented:
Looks great! Thanks to all the reviewers who weighed in, and sorry for the long delay.

mikepurvis commented:
Changes have been squashed on merge, so you'll need to rebase your local work.

CanyonTurtle commented:
@mikepurvis Thanks!! I will promptly rebase and follow up with more content - as I said, I have also ported the cutting-edge VEX V5 Robot Brain microcontroller, and I have tweaked this package slightly. Do you think it would be okay to put the new V5, and tweaks to this one, in one PR? or shall I split them up?

mikepurvis commented:
I'm easy-going on that kind of thing. If you have a lot stuff and/or expect to be iterating on this over time, would it make sense to simply add you as a maintainer of the rosserial repo so you can merge your own changes upon review?

CanyonTurtle commented:
@mikepurvis  I would appreciate that :) thanks!

That would be convenient, especially because I am about to open a PR for V5/updated cortex work. Just rebased my fork's jade-devel as you said, and it is ready to go.
